src/cfnlint/rules/resources/iam/Policy.py

```diff
@@ -97,60 +127,59 @@ def _check_policy_statement(self, branch, statement):
             message = 'IAM Policy statement missing Action or NotAction'
             matches.append(
                 RuleMatch(branch[:], message))
-        if 'Principal' in statement:
-            message = 'IAM Policy statement shouldn\'t have Principal'
-            matches.append(
-                RuleMatch(branch[:] + ['Principal'], message))
-        if 'NotPrincipal' in statement:
-            message = 'IAM Policy statement shouldn\'t have NotPrincipal'
-            matches.append(
-                RuleMatch(branch[:] + ['NotPrincipal'], message))
+        if is_identity_policy:
+            if 'Principal' in statement or 'NotPrincipal' in statement:
+                message = 'IAM Resource Policy statement shouldn\'t have Principal or NotPrincipal'
+                matches.append(
+                    RuleMatch(branch[:], message))
+        else:
+            if 'Principal' not in statement and 'NotPrincipal' not in statement:
+                message = 'IAM Resource Policy statement should have Principal or NotPrincipal'
+                matches.append(
+                    RuleMatch(branch[:] + ['Principal'], message))
         if 'Resource' not in statement and 'NotResource' not in statement:
             message = 'IAM Policy statement missing Resource or NotResource'
             matches.append(
                 RuleMatch(branch[:], message))
 
         return(matches)
 
-    def _check_policy(self, cfn, branch, policy):
-        """Checks a policy"""
+    def match_resource_properties(self, properties, resourcetype, path, cfn):
+        """Check CloudFormation Properties"""
         matches = list()
-        policy_document = policy.get('PolicyDocument', {})
-        matches.extend(
-            self._check_policy_document(
-                cfn, branch + ['PolicyDocument'], policy_document))
-
-        return matches
 
-    def match(self, cfn):
-        """Check IAM Policies Properties"""
+        is_identity_policy = True
+        if resourcetype in self.resources_and_keys:
+            is_identity_policy = False
 
-        matches = list()
+        key = None
+        if resourcetype in self.resources_and_keys:
+            key = self.resources_and_keys.get(resourcetype)
+        else:
```

Comment:
This seems dangerous. Should check against both type (resource_and_keys and idp_and_keys) and throw an exception if it's neither.